app/db/item_list_store.py

```diff
@@ -0,0 +1,224 @@
+
+        result = await cls.db.zadd(
+            cls.set_key, score=now() + cls.RETRY_DELAY, member=key
+        )
+        asyncio.create_task(notify_later())
+        return result
+
+    @classmethod
+    async def remove_item_list(cls, key: str) -> Any:
+        """
+        Remove a processed item list from the set.
+        """
+        result = await cls.db.zrem(cls.set_key, key)
+        return result
+
+    @classmethod
+    async def select_from_channel(cls) -> Optional[str]:
+        """
+        Wait until there's an item sent to the channel, then return it.
+
+        This takes advantage of the fact that the underlying connection
+        pool has free connections and will automatically use one of them
+        for the subscription.  That way we can be a subscriber while
+        other tasks are also being publishers.
+        """
+        try:
+            channels = await cls.db.subscribe(cls.item_list_channel_name)
+            key = await channels[0].get(encoding="utf-8")
+            await cls.db.unsubscribe(cls.item_list_channel_name)
+            return key
+        except RedisError:
+            log_error("Pub/Sub failed")
+            return None
+        except asyncio.CancelledError:
+            return None
+
+    @classmethod
+    async def select_for_processing(cls) -> Optional[str]:
+        """
+        Find the first item list that's ready for processing,
+        mark it as in-process, and return it.  The select
+        prioritizes any older item lists that were left from a prior run
+        over item lists that haven't been processed before.
+
+        Returns:
+            The item list key, if one is ready for processing, None otherwise.
+        """
+
+        async def mark_for_processing(key: str) -> bool:
+            """
+            Mark an item list key as being in process.
+            Returns whether setting the mark was successful.
+            """
+            new_score = now() + cls.IN_PROCESS
+            pipe = cls.db.multi_exec()
+            pipe.zadd(cls.set_key, score=new_score, member=key)
+            # return the exceptions rather than raising them because of
+            # this issue: https://github.com/aio-libs/aioredis/issues/558
+            values = await pipe.execute(return_exceptions=True)
+            return values[0] == 0
+
+        # because we are using optimistic locking, keep trying if we
+        # fail due to interference from other workers.
+        while True:
+            try:
+                start = now()
+                # optimistically lock the item set
+                await cls.db.watch(cls.set_key)
+                # first look for abandoned item lists from a prior run
+                item_lists = await cls.db.zrangebyscore(
+                    cls.set_key,
+                    min=start + (cls.RETRY_DELAY + cls.CLOCK_DRIFT),
+                    max=start + cls.IN_PROCESS - (cls.TIMEOUT + cls.CLOCK_DRIFT),
```

Comment:
why is this the value used for max?

Reply:
The query is looking for items left from a previous run.  These will be items that were marked for processing (had their score set to a week from now) but due to the crash were never completed and removed.  This means that they will have been marked for processing no more than TIMEOUT minutes ago, which means their max age will be one week from now minus the TIMEOUT (and adjusted for clock drift).